subprojects/dependency-management/src/main/java/org/gradle/internal/component/local/model/LocalVariantMetadata.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.component.local.model;
+
+
+import com.google.common.collect.ImmutableList;
+import org.gradle.api.artifacts.PublishArtifact;
+import org.gradle.api.artifacts.component.ComponentIdentifier;
+import org.gradle.api.internal.attributes.ImmutableAttributes;
+import org.gradle.internal.Describables;
+import org.gradle.internal.DisplayName;
+import org.gradle.internal.component.external.model.ImmutableCapabilities;
+import org.gradle.internal.component.model.DefaultVariantMetadata;
+import org.gradle.internal.model.CalculatedValueContainer;
+import org.gradle.internal.model.CalculatedValueContainerFactory;
+import org.gradle.internal.model.ModelContainer;
+
+import java.util.Collection;
+
+public class LocalVariantMetadata extends DefaultVariantMetadata {
```

Comment:
❕ Needs Javadoc for these new top-level classes.

Reply:
Also, should they be final?

Reply:
_should_ they be final? Perhaps. Though I don't very often make classes final I don't see a reason that they shouldn't be. I think making an inner class `static` is a lot more valuable, for example, than making a class `final`.

Reply:
For these specific classes: previously it would have been hard to extend them, as they were private and internal.  Now they could be extended, perhaps by accident, as there are a lot of similarly named classes in this subdomain, so it might be helpful to mark them as explicitly not meant for this.  

I'm not trying to make a general case for `final` here (though I do tend to lean that way).

Reply:
I've copied the javadoc from the follow-up PR to this PR.

Reply:
These are documented in the follow-up PR. I want to avoid conflicts with that PR, though it will be out for review shortly.